src/main.ts

```diff
@@ -1,409 +1,463 @@
 import * as YAML from "yaml"
 
+import Logger from "./logger"
+
 type Octokit = ReturnType<typeof github.getOctokit>
+type PR =
+  | Webhooks.PullRequestEvent["pull_request"]
+  | Webhooks.PullRequestReviewEvent["pull_request"]
 
-type ApprovalGroup = {
+type Rule = {
   name: string
   condition: string
-  check_type: "pr_diff" | "pr_files"
+  check_type: "diff" | "changed_files"
   min_approvals: number
-  users: Array<string> | undefined
-  teams: Array<string> | undefined
+  users: Array<string> | undefined | null
+  teams: Array<string> | undefined | null
 }
-const approvalGroupSchema = Joi.object<ApprovalGroup>().keys({
+const ruleSchema = Joi.object<Rule>().keys({
   name: Joi.string().required(),
   condition: Joi.string().required(),
-  check_type: Joi.string().valid("pr_diff", "pr_files").required(),
+  check_type: Joi.string().valid("diff", "changed_files").required(),
   min_approvals: Joi.number().required(),
-  users: Joi.array().items(Joi.string()),
-  teams: Joi.array().items(Joi.string()),
+  users: Joi.array().items(Joi.string()).optional().allow(null),
+  teams: Joi.array().items(Joi.string()).optional().allow(null),
 })
-type RulesConfiguration = {
-  approval_groups: ApprovalGroup[]
+type Configuration = {
+  rules: Rule[]
 }
-const rulesConfigurationSchema = Joi.object<RulesConfiguration>().keys({
-  approval_groups: Joi.array().items(approvalGroupSchema).required(),
+const configurationSchema = Joi.object<Configuration>().keys({
+  rules: Joi.array().items(ruleSchema).required(),
 })
 
-export function checkCondition(
-  check_type: string,
-  condition: RegExp,
-  pr_diff_body: { data: string },
-  pr_files_list: Set<string>,
-): boolean {
-  console.log(`###### BEGIN checkCondition ######`) //DEBUG
-  var condition_match: boolean = false
-  console.log(`condition: ${condition}`) //DEBUG
-  if (check_type === "pr_diff") {
-    if (pr_diff_body.data.match(condition) !== null) {
-      console.log(`Condition ${condition} matched`) //DEBUG
-      condition_match = true
-    }
-  }
-  if (check_type === "pr_files") {
-    for (const item of pr_files_list) {
-      if (item.match(condition)) {
-        console.log(`Condition ${condition} matched`) //DEBUG
-        condition_match = true
-      }
-    }
-  }
-  console.log(`###### END checkCondition ######`) //DEBUG
-  return condition_match
-}
+type RuleUserInfo = { team: string | null }
 
-export async function combineUsersTeams(
+const combineUsers = async function (
+  pr: PR,
   client: Octokit,
-  context: Context,
-  org: string,
-  pr_owner: string,
-  users: string[],
+  presetUsers: string[],
   teams: string[],
-): Promise<string[]> {
-  const full_approvers_list: Set<string> = new Set()
-  console.log(`###### BEGIN combineUsersTeams ######`) //DEBUG
-  console.log(`Users inside combine func: ${users} - `) //DEBUG
-  for (const user of users) {
-    if (pr_owner != user) {
-      console.log(`user: ${user}`) //DEBUG
-      full_approvers_list.add(user)
+) {
```

Comment:
I think it'd be useful to have an explicit return type. For example it's not obvious that it could return an instance of Error. 
Btw, is there a reason to return an error, but not to throw it?

Reply:
> I think it'd be useful to have an explicit return type.

True, I can add one here.

> For example it's not obvious that it could return an instance of Error.

IMO it's less obvious when any code can throw from anywhere at any time! Hence why I prefer `return Error("")` rather than `throw new Error("")`.

> Btw, is there a reason to return an error, but not to throw it?

I prefer to use exceptions only for **unrecoverable** and **unexpected** errors whenever possible and use `return new Error("")` when it is **recoverable** and **expected**. The principle is something along the lines of ["total functions"](https://ducmanhphan.github.io/2020-02-15-functional-programming-with-handling-errors/) and also [discriminated unions](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/discriminated-unions) from other languages. That sort of pattern is inconvenient in TypeScript so I generally don't use it everywhere.

Reply:
> I think it'd be useful to have an explicit return type

Added in https://github.com/paritytech/pr-custom-review/pull/27/commits/0109aa0aa62eb44fcd221d6ed972694d2f982114 but it should be noted that I explicitly prefer to not write the return type

https://github.com/paritytech/pr-custom-review/blob/2c4afa8463f9d96d17a73cc7071a9aafe42ce5a4/.eslintrc.js#L90

We don't have CI yet but I've been keeping the code linted and typechecking properly.

Reply:
Invalid response from a valid API sounds pretty unrecoverable to me.
Standard Javascript lib also don't mind using exceptions https://developer.mozilla.org/en-US/docs/Web/API/fetch#exceptions

Reply:
Changed to throw in 76aaafbdb11fac1c9093ef22ee206f6e9af66374